themes/default/content/blog/converting-full-terraform-programs-to-pulumi/index.md

```diff
@@ -0,0 +1,524 @@
+    <pulumi-choosable type="language" value="go"><code>notImplemented</code></pulumi-choosable>
+    <pulumi-choosable type="language" value="csharp"><code>NotImplemented</code></pulumi-choosable>
+</pulumi-chooser> function that will need to be filled in manually. For most projects, the converter should be able to convert 90-95% of the code without any TODOs, with only a small percentage of items to address manually, significantly reducing migration time compared to doing an entire migration by hand. We are actively improving the converter by adding support for missing features and improving the overall quality of the converted code to reduce the amount of manual fix-ups required.
+
+## Converting a Real World Program
+
+Let's walk through converting a Terraform codebase to Pulumi. [Avant Terraform Vault Setup](https://github.com/avantoss/vault-infra) is an open source project that provides a high-availability installation of Vault using a variety of managed AWS services. It defines a fairly complex installation with dozens of AWS resources in over 1,000 lines of Terraform HCL, including the main program and a Vault module. Let's convert it to Pulumi.
+
+First, clone the repo and `cd` into the directory containing the Terraform project:
+
+```bash
+$ git clone https://github.com/avantoss/vault-infra.git
+$ cd vault-infra/terraform/main
+```
+
+Next, run the converter:
+
+{{< chooser language "typescript,python,go,csharp" >}}
+{{% choosable language typescript %}}
+
+```bash
+$ pulumi convert --from terraform --language typescript --out pulumi
+```
+
+{{% /choosable %}}
+{{% choosable language python %}}
+
+```bash
+$ pulumi convert --from terraform --language python --out pulumi
+```
+
+{{% /choosable %}}
+{{% choosable language go %}}
+
+```bash
+$ pulumi convert --from terraform --language go --out pulumi
+```
+
+{{% /choosable %}}
+{{% choosable language csharp %}}
+
+```bash
+$ pulumi convert --from terraform --language csharp --out pulumi
+```
+
+{{% /choosable %}}
+{{< /chooser >}}
+
+![pulumi convert](pulumi-convert-from-terraform.gif)
+
+The converted code is generated in the specified `pulumi` output directory. There are two primary files generated from this program's conversion:
+
+* {{< langfile >}} contains the converted code for the main program
+* <pulumi-chooser type="language" options="typescript,python,go,csharp" option-style="none" class="inline">
+    <pulumi-choosable type="language" value="typescript"><code>vault.ts</code></pulumi-choosable>
+    <pulumi-choosable type="language" value="python"><code>vault.py</code></pulumi-choosable>
+    <pulumi-choosable type="language" value="go"><code>vault.go</code></pulumi-choosable>
+    <pulumi-choosable type="language" value="csharp"><code>vault.cs</code></pulumi-choosable>
+  </pulumi-chooser> contains the <code>Vault</code> Pulumi component, converted from the Terraform module
+
+### Addressing TODOs
```

Comment:
I think this is okay as is - but one thought is that we go into a lot of detail here - and possibly could move some of this into reference docs.  I could imagine having just one of these examples here, and then keeping the transition into "After addressing these and some other tweaks so that the code compiles."

This is especially true, as I imagine it will be fairly soon that we remove the need for there?

But possibly easy enough to just leave as is, and remove them over the coming weeks as we address the underlying gaps?

Reply:
> This is especially true, as I imagine it will be fairly soon that we remove the need for there?

Yes! I'm pretty sure fixing "replace" is just a case of getting a new std release and telling the converter to use it, so will probably be in next week (Maybe even for the main release on Monday)

Reply:
Moved to just a single example for the blog post. I'll save the rest for a more detailed walkthrough in reference docs.